File: ndai/api/schemas/secrets.py
"""Pydantic schemas for Conditional Recall."""

from datetime import datetime

from pydantic import BaseModel, Field


class SecretPolicy(BaseModel):
    allowed_actions: list[str] = Field(..., min_length=1)
    max_uses: int = Field(..., gt=0)


class SecretCreateRequest(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    description: str | None = None
    secret_value: str = Field(..., min_length=1)
    policy: SecretPolicy


class SecretResponse(BaseModel):
    id: str
    name: str
    description: str | None
    policy: dict
    uses_remaining: int
    status: str
    created_at: datetime
    is_owner: bool = False


class SecretUseRequest(BaseModel):
    action: str = Field(..., min_length=1)


class SecretUseResponse(BaseModel):
    action: str
    result: str
    success: bool
    secret_name: str
    attestation_available: bool = True
    policy_report: dict | None = None
    policy_constraints: list[dict] | None = None
    egress_log: list[dict] | None = None
    verification: dict | None = None


class SecretAccessLogResponse(BaseModel):
    id: int
    secret_id: str
    requester_id: str
    requester_display_name: str | None = None
    action_requested: str
    status: str
    result_summary: str | None
    verification_data: dict | None = None
    created_at: datetime

File: frontend/src/api/secrets.ts
import { get, post } from "./client";

export interface SecretPolicy {
  allowed_actions: string[];
  max_uses: number;
}

export interface SecretCreateRequest {
  name: string;
  description?: string;
  secret_value: string;
  policy: SecretPolicy;
}

export interface SecretResponse {
  id: string;
  name: string;
  description: string | null;
  policy: SecretPolicy;
  uses_remaining: number;
  status: string;
  created_at: string;
  is_owner: boolean;
}

export interface SecretUseResponse {
  action: string;
  result: string;
  success: boolean;
  secret_name: string;
  attestation_available: boolean;
  policy_report: any | null;
  policy_constraints: any[] | null;
  egress_log: any[] | null;
  verification: any | null;
}

export interface AccessLogEntry {
  id: number;
  secret_id: string;
  requester_id: string;
  requester_display_name: string | null;
  action_requested: string;
  status: string;
  result_summary: string | null;
  verification_data: any | null;
  created_at: string;
}

export const createSecret = (data: SecretCreateRequest) => post<SecretResponse>("/secrets/", data);
export const listMySecrets = () => get<SecretResponse[]>("/secrets/");
export const listAvailableSecrets = () => get<SecretResponse[]>("/secrets/available");
export const getSecret = (id: string) => get<SecretResponse>(`/secrets/${id}`);
export const useSecret = (id: string, action: string) => post<SecretUseResponse>(`/secrets/${id}/use`, { action });
export const revokeSecret = (id: string) => post<SecretResponse>(`/secrets/${id}/revoke`, {});
export const getAccessLog = (id: string) => get<AccessLogEntry[]>(`/secrets/${id}/access-log`);
